src/blockstore/meta.zig

```diff
@@ -0,0 +1,344 @@
+const std = @import("std");
+const sig = @import("../lib.zig");
+
+const Allocator = std.mem.Allocator;
+
+const BitFlags = sig.utils.bitflags.BitFlags;
+const Shred = sig.shred_collector.shred.Shred;
+const CodingShred = sig.shred_collector.shred.CodingShred;
+const Slot = sig.core.Slot;
+const SortedSet = sig.utils.collections.SortedSet;
+
+/// The Meta column family
+pub const SlotMeta = struct {
+    // allocator: Allocator,
+    /// The number of slots above the root (the genesis block). The first
+    /// slot has slot 0.
+    slot: Slot,
+    /// The total number of consecutive shreds starting from index 0 we have received for this slot.
+    /// At the same time, it is also an index of the first missing shred for this slot, while the
+    /// slot is incomplete.
+    consumed: u64,
+    /// The index *plus one* of the highest shred received for this slot.  Useful
+    /// for checking if the slot has received any shreds yet, and to calculate the
+    /// range where there is one or more holes: `(consumed..received)`.
+    received: u64,
+    /// The timestamp of the first time a shred was added for this slot
+    first_shred_timestamp: u64,
+    /// The index of the shred that is flagged as the last shred for this slot.
+    /// None until the shred with LAST_SHRED_IN_SLOT flag is received.
+    last_index: ?u64,
+    /// The slot height of the block this one derives from.
+    /// The parent slot of the head of a detached chain of slots is None.
+    parent_slot: ?Slot,
+    /// The list of slots, each of which contains a block that derives
+    /// from this one.
+    next_slots: std.ArrayList(Slot),
+    /// Connected status flags of this slot
+    connected_flags: ConnectedFlags,
+    /// Shreds indices which are marked data complete.  That is, those that have the
+    /// [`ShredFlags::DATA_COMPLETE_SHRED`][`crate::shred::ShredFlags::DATA_COMPLETE_SHRED`] set.
+    completed_data_indexes: SortedSet(u32),
+
+    const Self = @This();
+
+    pub fn init(allocator: std.mem.Allocator, slot: Slot, parent_slot: ?Slot) Self {
+        const connected_flags = if (slot == 0)
+            // Slot 0 is the start, mark it as having its' parent connected
+            // such that slot 0 becoming full will be updated as connected
+            ConnectedFlags.from(.parent_connected)
+        else
+            ConnectedFlags{};
+        return .{
+            // .allocator = allocator,
+            .slot = slot,
+            .parent_slot = parent_slot,
+            .connected_flags = connected_flags,
+            .consumed = 0,
+            .received = 0,
+            .first_shred_timestamp = 0,
+            .last_index = null,
+            .next_slots = std.ArrayList(Slot).init(allocator),
+            .completed_data_indexes = SortedSet(u32).init(allocator),
+        };
+    }
+
+    pub fn deinit(self: Self) void {
+        self.next_slots.deinit();
+    }
+
+    // TODO copy on write
```

Comment:
this TODO still valid?

Reply:
I think I put it there because it was relevant for a specific usage of clone being particularly wasteful. But I must have removed that usage.

cd8719524603b05de51a16d5f57f8a7d77a6b21e